onnx/defs/tensor/defs.cc

```diff
@@ -1797,62 +1797,115 @@ ONNX_OPERATOR_SET_SCHEMA(
-          // depends on actual input data, but the rank is always 1
-          ctx.getOutputType(0)
-              ->mutable_tensor_type()
-              ->mutable_shape()
-              ->add_dim();
+          const TypeProto* xTensorProto = ctx.getInputType(0);
+          TypeProto* yTensorProto = ctx.getOutputType(0);
+          TypeProto* indicesTensorProto = nullptr;
+          TypeProto* inverseIndicesTensorProto = nullptr;
+          TypeProto* countsTensorProto = nullptr;
+          
+          auto num_outputs = ctx.getNumOutputs();
+          if (num_outputs >= 2) {
+            indicesTensorProto = ctx.getOutputType(1);
+            updateOutputElemType(ctx, 1, TensorProto::INT64);
+          }
 
-          ctx.getOutputType(2)
-              ->mutable_tensor_type()
-              ->mutable_shape()
-              ->add_dim();
+          if (num_outputs >= 3) {
+            inverseIndicesTensorProto = ctx.getOutputType(2);
+            updateOutputElemType(ctx, 2, TensorProto::INT64);
+          }
 
-          // if the input shape doesn't exist, further shape inference is not possible
-          if (!hasNInputShapes(ctx, 1)) {
-            return;
+          if (num_outputs >= 4) {
+            countsTensorProto = ctx.getOutputType(3);
+            updateOutputElemType(ctx, 3, TensorProto::INT64);
           }
 
-          // 'idx' output has same shape as input
-          propagateShapeFromInputToOutput(ctx, 0, 1);
+          auto axisAttr = ctx.getAttribute("axis");
+          if (!axisAttr) {
+            // Flatten case.
+            // 'Y', 'indices', and 'counts' are 1-D tensors of unknown dimension.
+            yTensorProto->mutable_tensor_type()->mutable_shape()->add_dim();
+            if (indicesTensorProto)
+              indicesTensorProto->mutable_tensor_type()->mutable_shape()->add_dim();
+            if (countsTensorProto)
+              countsTensorProto->mutable_tensor_type()->mutable_shape()->add_dim();
+
+            // 'inverse_indices' has the same shape as 'X'.
+            if (inverseIndicesTensorProto)
+              propagateShapeFromInputToOutput(ctx, 0, 2);
+          }
+          else 
+          {
+            // Axis is provided.
+            int axis = static_cast<int>(axisAttr->i());
+
+            // 'Y' has the same dimensions as 'X' except the dimension 'axis' which is unknown.
+            const TensorShapeProto& input_shape = xTensorProto->tensor_type().shape();
+
+            for (int i = 0; i < input_shape.dim_size(); i++) {
+              auto* dim = yTensorProto->mutable_tensor_type()->mutable_shape()->add_dim();
+              if (i != axis)
+                dim->set_dim_value(input_shape.dim(i).dim_value());
+            }
 
-          return;
-        }));
+            // 'indices', 'inverse_indices', and 'counts' are 1-D tensors of unknown dimension.
+            if (indicesTensorProto)
+              indicesTensorProto->mutable_tensor_type()->mutable_shape()->add_dim();
```

Comment:
According to the description earlier, indices has same shape as input.

Reply:
Similar comments for the other outputs as well. I assume that they will not be flattened if axis is specified?

Reply:
Updated the spec to take output shapes of numpy. indices, inverse_indices, and counts are all 1-D tensors.